kubectl delete resource : remove all resources

sdminonne opened the pull request:
To fix issue https://github.com/GoogleCloudPlatform/kubernetes/issues/4380

sdminonne committed:
kubectl delete resource : remove all resources

smarterclayton commented on pkg/kubectl/cmd/delete.go:
```diff
@@ -69,10 +69,14 @@ Examples:
 				Do()
-			checkErr(r.Err())
+
+			if r.Err() != nil {
```
Hrm - this change means that the contract of resource builder changes (if it returns an error, you *cannot* use the visited responses).  What was it for?

sdminonne replied:
my understanding is that checkErr is calling glog.FatalDepth and it's not returning on cmd.out. This prevents unit test to verify the output of the command.

smarterclayton replied:
Ah - this is a more general problem with our client tools.  Since we have this problem everywhere in kubectl (we can't test failures) I'd probably prefer to use `checkErr` for here and do the test in hack/test-cmd.sh (which can test failure).  We do need to correct that problem in kubectl eventually, but I'm a bit worried about having two different ways of checking errors start creeping into the code.

sdminonne replied:
OK. I'll remove the unit test and I'll add a test to hack/test-cmd.sh

smarterclayton replied:
Don't be afraid to add lots more tests to test-cmd.sh - it's really our best "integrated" test framework for API calls and CLI behavior.

----- Original Message -----
> >  				Flatten().
> >  				Do()
> > -			checkErr(r.Err())
> > +
> > +			if r.Err() != nil {
> 
> OK. I'll remove the unit test and I'll add a test to hack/test-cmd.sh
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/GoogleCloudPlatform/kubernetes/pull/4381/files#r24587817

smarterclayton commented on pkg/kubectl/resource/builder.go:
```diff
@@ -224,6 +224,28 @@ func (b *Builder) ResourceTypeOrNameArgs(args ...string) *Builder {
+// retrieved will be retrieved from the server. The resources retrieved must be restricted via
+// a label selector othrerwise an error is reported. If more that two arguments are provided an
+// error is set.
+func (b *Builder) SelectedResourceTypeOrNameArgs(args ...string) *Builder {
```
I think the name is slightly confusing - it wasn't clear to me the distinction at first.  Maybe we should add a boolean argument to ResourceTypeOrNameArgs "allowEmptySelector"?

sdminonne replied:
OK

smarterclayton commented:
Also, maybe we should have a "select all" flag like `-a` that can be used instead of `-l`?  I assume the goal here is to allow `kubectl delete foo,bar -s zone=1`, but what if I really do want to delete everything?  `kubectl delete foo,bar -a`?

I'm definitely ok with requiring a user to clarify delete all (since it's dangerous)

sdminonne commented:
I would vote for a `-all`...

sdminonne commented:
@smarterclayton
I've added --all  (-a already registered)
I've removed the misleading and the useless unit test
Added the allowEmptySelector 
Some testcase in cmd/test-hack.cmd added.

smarterclayton commented:
Hrm, gendocs is flagging you on travis

sdminonne commented:
Oups... let me redo this

lavalamp commented:
This looks pretty good... I can't seem to find a CLA from you, though. Have you filed one with us? See CONTRIBUTING.md

sdminonne commented:
My company (amadeus) signed the CLA.

smarterclayton commented:
LGTM

lavalamp commented:
@sdminonne can you go to https://cla.developers.google.com/ and see if your CLA is listed?

sdminonne commented:
@lavalamp  I can't find it.  But I do know that our lawyers signed the CLA. On Monday I'll come back with more info: lawyers don't work during the week-end. :)

lavalamp commented:
@sdminonne thanks for looking into it, sorry for the inconvenience. I'm working on enabling a CLA checking bot so we'll catch this right away next time.

schuhmi2 commented:
@sdminonne lawyers don't, but the open source office unofficially does.
@lavalamp I can confirm that he is covered under our corporate CLA for Amadeus SAS, But only with use of the Google bot, as our CLA only lists the Google group, not individual users. My corporate email is in my profile if you need further official confirmation.

sdminonne commented:
@schuhmi2 the same for Eric https://github.com/GoogleCloudPlatform/kubernetes/pull/4436

smarterclayton commented:
Will merge on Monday.

sdminonne commented:
@smarterclayton thanks

smarterclayton commented:
Oops, forgot yesterday.

smarterclayton commented:
Rerunning travis and will merge.

sdminonne commented:
@smarterclayton no problem :)
